# Introduction to the Python language


We introduce here the Python language. Only the bare minimum
necessary for getting started with Numpy and Scipy is addressed here.
To learn more about the language, consider going through the
excellent tutorial http://www.scipy-lectures.org/intro/index.html. Dedicated books
are also available, such as http://www.diveintopython.net/.

Python is a **programming language**, as are C, Fortran, BASIC, PHP,
etc. Some specific features of Python are as follows:

* an *interpreted* (as opposed to *compiled*) language. Contrary to e.g.
C or Fortran, one does not compile Python code before executing it. In
addition, Python can be used **interactively**: many Python
interpreters are available, from which commands and scripts can be
executed.

* a free software released under an **open-source** license: Python can
be used and distributed free of charge, even for building commercial
software.

* **multi-platform**: Python is available for all major operating
systems, Windows, Linux/Unix, MacOS X, most likely your mobile phone
OS, etc.

* a very readable language with clear non-verbose syntax

* a language for which a large variety of high-quality packages are
available for various applications, from web frameworks to scientific
computing.

* a language very easy to interface with other languages, in particular C
and C++.

* Some other features of the language are illustrated just below. For
example, Python is an object-oriented language, with dynamic typing
(the same variable can contain objects of different types during the
course of a program).


See https://www.python.org/about/ for more information about
distinguishing features of Python.

# Hello world

you type in the cells, execute commands with shift + Enter

```print("Hello world")```

In [ ]:
print("Hello world")

# Basic types

## Integers

In [ ]:
1+1

We can assign values to variables with `=`

    a = 4
    type(a)

In [ ]:
a = 4

In [ ]:
type(a)

Note that one does not declare the type of a variable before assigning its value. In C, conversely, one should write:

```int a = 4;```

## Floats
There exists a floating point type that is created when the variable has decimal values:

    c = 2.1
    type(c)

In [ ]:
c = 2.1

In [ ]:
type(c)

## Boolean
Similarly, boolean types are created from a comparison:

In [ ]:
3 > 4

    test = (3 > 4)
    type(test)

In [ ]:
test = (3 > 4)

In [ ]:
type(test)

_Tip_:A Python shell can therefore replace your pocket calculator, with the basic arithmetic operations ``+``, ``-``, ``*``, ``/``, ``%`` (modulo) natively implemented

## Type conversion (casting)

In [ ]:
a = 1
type(a)

In [ ]:
b = float(a)
type(b)

### Exercise <^^>: Compute the floating point expression of 3 / 2

#  Containers

_Tip_: Python provides many efficient types of containers, in which collections of objects can be stored.

## Lists

A list is an ordered collection of objects, that may have different types. For example:


    colors = ['red', 'blue', 'green', 'black', 'white']
    type(colors)

In [ ]:
colors = ['red', 'blue', 'green', 'black', 'white']

In [ ]:
type(colors)

Indexing: accessing individual objects contained in the list::

In [ ]:
colors[2]

<h3 style="color: red"> WARNING: **Indexing starts at 0** (as in C), not at 1 (as in Fortran or Matlab)!</h3>

Counting from the end with negative indices:

In [ ]:
colors[-2]

### <^^> Exercise: get the before last element in colors

## Slicing: obtaining sublists of regularly-spaced elements


In [ ]:
colors

In [ ]:
colors[1:5:2]

**Slicing syntax**: ``colors[start:stop:stride]``

  _Tip_: All slicing parameters are optional::

In [ ]:
colors

In [ ]:
colors[3:]

In [ ]:
colors[:3]

### <^^> Exercise: what is the output of ```colors[::2]```. Why?



In [ ]:
colors[::2]

### <^^> Exercise: reverse a list

In [ ]:
colors[::-1]


# Strings

Different string syntaxes (simple, double or triple quotes)::

    s = 'Hello, how are you?'
    s = "Hi, what's up"
    s = '''Hello,                 # tripling the quotes allows the
           how are you'''         # the string to span more than one line
    s = """Hi,
    what's up?"""

In [ ]:
s = 'asdfdfdf'
type(s)

In [ ]:
s = "asdfdfd"

In [ ]:
s = """asdfadsf
asdfdf
asdfdf
"""

# Dictionaries

A dictionary is basically an efficient table that **maps keys to values**. It is an **unordered** container

In [ ]:
tel = {'emmanuelle': 5752, 'sebastian': 5578}

In [2]:
tel

{'emanuelle': 5752, 'sebastian': 5578}

In [3]:
tel['francis']

KeyError: 'francis'

In [4]:
tel['francis'] = 5919

In [5]:
tel

{'emanuelle': 5752, 'francis': 5919, 'sebastian': 5578}

In [6]:
tel['sebastian']

5578

In [7]:
tel.keys()

dict_keys(['sebastian', 'emanuelle', 'francis'])

In [8]:
tel.values()

dict_values([5578, 5752, 5919])

In [9]:
'francis' in tel

True

It can be used to conveniently store and retrieve values
associated with a name (a string for a date, a name, etc.). See
https://docs.python.org/tutorial/datastructures.html#dictionaries
for more information.

_Tip_: A dictionary can have keys (resp. values) with different types:

In [11]:
d = {'a':1, 'b':2, 3: 'asdf'}
d

{'a': 1, 3: 'asdf', 'b': 2}

In [12]:
del d['a']

In [15]:
d

{3: 'asdf', 'b': 2}

### <^^> Exercise: assign the telephone of sebastian to francis, and that of francis to sebastian.


In [22]:
tel = {'emanuelle': 5752, 'sebastian' : 5578, 'francis' : 1234}

In [23]:
tmp = tel['francis']
tel['francis'] = tel['sebastian']
tel['sebastian'] = tmp

In [24]:
tel

{'emanuelle': 5752, 'francis': 5578, 'sebastian': 1234}

# Control Flow

Controls the order in which the code is executed.

## if/elif/else

    if 2**2 == 4:
        print('Obvious!')

In [29]:
if 2 ** 2 == 5:
   print('Obvious')
print('YES')

YES



## Blocks are delimited by indentation!

_Tip_: Type the following lines in your Python interpreter, and be careful
to **respect the indentation depth**. The Ipython shell automatically
increases the indentation depth after a colon ``:`` sign; to
decrease the indentation depth, go four spaces to the left with the
Backspace key. Press the Enter key twice to leave the logical block.

    >>> a = 10

    >>> if a == 1:
    ...     print(1)
    ... elif a == 2:
    ...     print(2)
    ... else:
    ...     print('A lot')
    A lot



In [31]:
a = 10
if a == 1:
    print(1)
elif a == 2:
    print(2)

# for/range

Iterating with an index::

    >>> for i in range(4):
    ...     print(i)
    0
    1
    2
    3


In [32]:
for i in range(4):
    print(i + 1)

1
2
3
4


In [39]:
a = []


In [37]:
for i in range(0, 101, 2):
    print(i)

0
2
4
6
8
10
12
14
16
18
20
22
24
26
28
30
32
34
36
38
40
42
44
46
48
50
52
54
56
58
60
62
64
66
68
70
72
74
76
78
80
82
84
86
88
90
92
94
96
98
100


### <^^> Exercise: print all even numbers between 0 and 100

But most often, it is more readable to iterate over values::

    >>> for word in ('cool', 'powerful', 'readable'):
    ...     print('Python is %s' % word)
    Python is cool
    Python is powerful
    Python is readable

In [40]:
for word in ['green', 'blue', 'yellow']:
    print(word)

green
blue
yellow



### List Comprehensions


    >>> [i**2 for i in range(4)]
    [0, 1, 4, 9]


In [41]:
[2 * i for i in range(51)]

[0,
 2,
 4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 94,
 96,
 98,
 100]

In [42]:
a = 10
b = 1
while b < a:
    b = b + 1
    print(b)

2
3
4
5
6
7
8
9
10


### <^^> Exercise

Compute the decimals of Pi using the Wallis formula:

$$
\pi = 2 \prod_{i=1}^{100} \frac{4i^2}{4i^2 - 1}
$$


In [50]:
pi = 2
eps = 1e-6
dif = 2 * eps
i = 1
while(dif > eps):
    old_pi = pi
    pi *= 4 * i ** 2 / (4 * i ** 2 - 1)
    dif = pi - old_pi
    i += 1

In [51]:
pi

3.140707822589477



### while/break/continue

Typical C-style while loop (Mandelbrot problem)::

    >>> z = 1 + 1j
    >>> while abs(z) < 100:
    ...     z = z**2 + 1
    >>> z
    (-134+352j)



In [43]:
[i while i < 10; i += 1]

SyntaxError: invalid syntax (<ipython-input-43-d058ae05126b>, line 1)

# Defining functions

### Function definition
    def test():
        print('in test function')

    >>> test()
    in test function

Warning: Function blocks must be indented as other control-flow blocks.



In [52]:
def test():
    print('in test function')

In [53]:
test()

in test function


### Return statement

Functions can *optionally* return values.


    >>> def disk_area(radius):
       ...:     return 3.14 * radius * radius
       ...:

    >>> disk_area(1.5)
    >>> 7.0649999999999995

Note: By default, functions return ``None``.

The syntax to define a function:

    * the ``def`` keyword;

    * is followed by the function's **name**, then

    * the arguments of the function are given between parentheses followed
      by a colon.

    * the function body;

    * and ``return object`` for optionally returning values.





### Parameters

Mandatory parameters (positional arguments)


    >>> def double_it(x):
       ....:     return x * 2
       ....:

    >>> double_it(3)
    >>> 6

    >>> double_it()
    ---------------------------------------------------------------------------
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: double_it() takes exactly 1 argument (0 given)

Optional parameters (keyword or named arguments)


    >>> def double_it(x=2):
       ....:     return x * 2
       ....:

    >>> double_it()
    >>> 4

    >>> double_it(3)
    >>> 6

Keyword arguments allow you to specify *default values*.


In [54]:
def square(x):
    return x ** 2

In [56]:
square(4)

16

### Exercise: compute the 20-th Fibonacci number.

# Object-oriented programming (OOP)

Python supports object-oriented programming (OOP). The goals of OOP are:

    * to organize the code, and

    * to re-use code in similar contexts.



Here is a small example: we create a Student *class*, which is an object
gathering several custom functions (*methods*) and variables (*attributes*),
we will be able to use::

    >>> class Student(object):
    ...     def __init__(self, name):
    ...         self.name = name
    ...     def set_age(self, age):
    ...         self.age = age
    ...     def set_major(self, major):
    ...         self.major = major
    ...
    >>> anna = Student('anna')
    >>> anna.set_age(21)
    >>> anna.set_major('physics')

In the previous example, the Student class has ``__init__``, ``set_age`` and
``set_major`` methods. Its attributes are ``name``, ``age`` and ``major``. We
can call these methods and attributes with the following notation:
``classinstance.method`` or  ``classinstance.attribute``.  The ``__init__``
constructor is a special method we call with: ``MyClass(init parameters if
any)``.

Now, suppose we want to create a new class MasterStudent with the same
methods and attributes as the previous one, but with an additional
``internship`` attribute. We won't copy the previous class, but
**inherit** from it::

    >>> class MasterStudent(Student):
    ...     internship = 'mandatory, from March to June'
    ...
    >>> james = MasterStudent('james')
    >>> james.internship
    'mandatory, from March to June'
    >>> james.set_age(23)
    >>> james.age
    23

In [60]:
class Numbers:
    def add(self, x, y):
        return x + y
    def substract(self, x, y):
        return x - y

In [61]:
a = Numbers()

In [62]:
a.add(2, 3)

5

### Exercise: define a class for an integer number with methods ```add``` and ```substract``` 